lib/smack/SplitStructLoadStore.cpp

```diff
@@ -0,0 +1,151 @@
+#include "smack/SplitStructLoadStore.h"
+#include "llvm/IR/InstIterator.h"
+
+namespace smack {
+
+using namespace llvm;
+
+std::vector<Value*> getFirsts(std::vector<std::pair<Value*, unsigned> > lst)
+{
+    std::vector<Value*> ret;
+    for (auto p = lst.begin(); p != lst.end(); ++p)
+      ret.push_back(std::get<0>(*p));
+    return ret;
+}
+
+std::vector<unsigned> getSeconds(std::vector<std::pair<Value*, unsigned> > lst)
+{
+    std::vector<unsigned> ret;
+    for (auto p = lst.begin()+1; p != lst.end(); ++p)
+      ret.push_back(std::get<1>(*p));
+    return ret;
+}
+
+bool SplitStructLoadStore::runOnModule(Module& M)
+{
+  DL = &M.getDataLayout();
+
+  for (auto& F : M) {
+    for(inst_iterator I = inst_begin(F), E = inst_end(F); I != E; ++I) {
+      if (LoadInst* li = dyn_cast<LoadInst>(&*I)) {
+        bool isReturnValue = false;
+        for (auto u = li->user_begin(); u != li->user_end(); ++u)
+        {
+          if (auto ri = dyn_cast<ReturnInst>(*u)) {
+            isReturnValue = true;
+            break;
+          }
+        }
+        if (!isReturnValue && li->getType()->isAggregateType()) {
+          splitStructLoad(li);
+        }
+      } else if (StoreInst* si = dyn_cast<StoreInst>(&*I)) {
+        Value* P = si->getPointerOperand();
+        Value* V = si->getOperand(0)->stripPointerCasts();
+        if (V->getType()->isAggregateType()) {
+          splitStructStore(si, P, V);
+        }
+      }
+    }
+  }
+
+  for (auto& i : toRemove)
+    i->eraseFromParent();
+  return true;
+}
+
+void SplitStructLoadStore::splitStructLoad(LoadInst* li)
+{
+  if (StructType* st = dyn_cast<StructType>(li->getType())) {
+    std::vector<std::pair<Value*, unsigned> > idx;
+    IRBuilder<> irb(li);
+    li->replaceAllUsesWith(buildStructs(&irb, li->getPointerOperand(), st, nullptr, idx));
+    toRemove.push_back(li);
```

Comment:
This could just as easily be done at the call site of `splitStructLoad`, making `toRemove` a local variable.